launch/launch/utilities/signal_management.py

```diff
@@ -12,133 +12,102 @@
-            __custom_sigterm_handler(signum, frame, __original_sigterm_handler)
-        elif callable(__original_sigterm_handler):
-            __original_sigterm_handler(signum, frame)
-
-    # signals must be registered in the main thread, but print a nicer message if we're not there
-    try:
-        signal.signal(signal.SIGINT, __on_sigint)
-        signal.signal(signal.SIGTERM, __on_sigterm)
-        if platform.system() != 'Windows':
-            # Windows does not support SIGQUIT
-            signal.signal(signal.SIGQUIT, __on_sigquit)
-    except ValueError:
-        logger = launch.logging.get_logger(__name__)
-        logger.error("failed to set signal handlers in '{}'".format(__name__))
-        logger.error('this function must be called in the main thread')
-        raise
+    def __init__(
+        self,
+        loop: asyncio.AbstractEventLoop
+    ):
+        """
+        Instantiate manager.
+
+        :param loop: event loop that will handle the signals.
+        """
+        self.__loop = loop
+        self.__handlers = {}
+        self.__prev_wsock_fd = -1
+        self.__wsock, self.__rsock = socket.socketpair()
+        self.__wsock.setblocking(False)
+        self.__rsock.setblocking(False)
+
+    def __enter__(self):
+        self.__loop.add_reader(self.__rsock.fileno(), self.__handle_signal)
+        self.__prev_wsock_fd = signal.set_wakeup_fd(self.__wsock.fileno())
+        return self
+
+    def __exit__(self, type_, value, traceback):
+        assert self.__wsock.fileno() == signal.set_wakeup_fd(self.__prev_wsock_fd)
+        self.__loop.remove_reader(self.__rsock.fileno())
+
+    def __handle_signal(self):
+        while True:
+            try:
+                data = self.__rsock.recv(4096)
+                if not data:
+                    break
+                for signum in data:
+                    if signum not in self.__handlers:
+                        continue
+                    self.__handlers[signum](signum)
+                if self.__prev_wsock_fd != -1:
+                    os.write(self.__prev_wsock_fd, data)
+            except InterruptedError:
+                continue
+            except BlockingIOError:
+                break
+
+    def handle(
+        self,
+        signum: Union[signal.Signals, int],
+        handler: Optional[Callable[[int], None]],
+    ) -> Optional[Callable[[int], None]]:
+        """
+        Register a callback for asynchronous handling of a given signal.
+
+        :param signum: number of the signal to be handled
+        :param handler: callback taking a signal number
+          as its sole argument, or None
+        :return: previous handler if any, otherwise None
+        """
+        signum = signal.Signals(signum)
+        if handler is not None and not callable(handler):
+            raise ValueError('signal handler must be a callable')
+        old_handler = self.__handlers.get(signum, None)
+        self.__handlers[signum] = handler
```

Comment:
IIUC `handler` can be `None` here, which can be problematic with the above:

```python3
                    if signum not in self.__handlers:
                        continue
                    self.__handlers[signum](signum)
```

maybe the correct code is
```suggestion
        if handler is not None:
          self.__handlers[signum] = handler
        else:
          self.__handlers.pop(signum, None)
```

Reply:
Oof, yes. See 3c753e60b1d4a3c16269b4aeb257287c17b632cc. I extended the test to cover this.